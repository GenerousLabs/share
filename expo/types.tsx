export type BottomTabParamList = {
  TabOne: undefined;
  TabTwo: undefined;
};

export type TabOneParamList = {
  TabOneScreen: undefined;
};

export type TabTwoParamList = {
  TabTwoScreen: undefined;
};
