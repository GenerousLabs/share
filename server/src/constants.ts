import { join } from "path";

export const __DEV__ = process.env.NODE_ENV === "production";

// NOTE: We need to run `process.cwd()` here, when running it inside `join()` we
// get the `dist/` folder. No idea why. Insanity.
export const CWD = process.cwd();
export const DATA_PATH = join(CWD, "data");

export const POSTOFFICE_PATH = join(DATA_PATH, "postoffice");

export const REPOS_ROOT = join(DATA_PATH, "repos");
export const COMMANDS_REPO_NAME = "commands.git" as const;

export const REPO_TEMPLATE_PATH = join(CWD, "templates/empty-repo");

export const USER_TOKEN_PATH = join(
  REPOS_ROOT,
  "commands.git",
  "user_tokens.yaml"
);

export const ENOENT = "ENOENT" as const;

// The ID is comprised of 3 words joined with a dash, not sure what the minimum
// length of each word is though.
export const MINIMUM_POSTOFFICE_ID_LENGTH = 12;

// NOTE: These 3 are copied to `expo/src/shared.constants.ts` file and must be
// manually kept in sync in both.
export const READ_TOKENS_FILE_NAME = "read_tokens.txt" as const;
export const MINIMUM_USERNAME_LENGTH = __DEV__ ? 1 : 3;
export const MINIMUM_TOKEN_LENGTH = __DEV__ ? 1 : 20;
