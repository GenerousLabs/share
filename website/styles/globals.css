/* montserrat-regular - latin */
@font-face {
  font-family: "Montserrat";
  font-style: normal;
  font-weight: 400;
  src: url("/_static/fonts/montserrat-v15-latin-regular.eot"); /* IE9 Compat Modes */
  src: local(""),
    url("/_static/fonts/montserrat-v15-latin-regular.eot?#iefix")
      format("embedded-opentype"),
    /* IE6-IE8 */ url("/_static/fonts/montserrat-v15-latin-regular.woff2")
      format("woff2"),
    /* Super Modern Browsers */
      url("/_static/fonts/montserrat-v15-latin-regular.woff") format("woff"),
    /* Modern Browsers */ url("/_static/fonts/montserrat-v15-latin-regular.ttf")
      format("truetype"),
    /* Safari, Android, iOS */
      url("/_static/fonts/montserrat-v15-latin-regular.svg#Montserrat")
      format("svg"); /* Legacy iOS */
}

/* montserrat-700 - latin */
@font-face {
  font-family: "Montserrat";
  font-style: normal;
  font-weight: 700;
  src: url("/_static/fonts/montserrat-v15-latin-700.eot"); /* IE9 Compat Modes */
  src: local(""),
    url("/_static/fonts/montserrat-v15-latin-700.eot?#iefix")
      format("embedded-opentype"),
    /* IE6-IE8 */ url("/_static/fonts/montserrat-v15-latin-700.woff2")
      format("woff2"),
    /* Super Modern Browsers */
      url("/_static/fonts/montserrat-v15-latin-700.woff") format("woff"),
    /* Modern Browsers */ url("/_static/fonts/montserrat-v15-latin-700.ttf")
      format("truetype"),
    /* Safari, Android, iOS */
      url("/_static/fonts/montserrat-v15-latin-700.svg#Montserrat")
      format("svg"); /* Legacy iOS */
}

html,
body {
  padding: 0;
  margin: 0;
  font-family: -apple-system, Montserrat, BlinkMacSystemFont, Segoe UI, Roboto,
    Oxygen, Ubuntu, Cantarell, Fira Sans, Droid Sans, Helvetica Neue, sans-serif;
}

a {
  color: inherit;
  text-decoration: underline;
  font-weight: bold;
}

p {
  font-size: 16px;
}

* {
  box-sizing: border-box;
}

.appWrapper {
  display: flex;
  flex-direction: column;
  align-items: center;
  min-height: 100vh;
}

@media screen and (max-width: 580px) {
  .appContainer {
    max-width: 580px;
    flex-grow: 1;
    padding: 15px;
  }
}
@media screen and (min-width: 581px) {
  .appContainer {
    width: 580px;
    flex-grow: 1;
  }
}

header {
  position: sticky;
  width: 100vw;
  top: 0;
  background-color: #212121;
  color: white;
  display: flex;
  justify-content: center;
  margin-bottom: 20px;
}
header h1 {
  text-transform: uppercase;
  font-size: 26px;
}

.logoContainer {
  display: flex;
  justify-content: flex-start;
}

h2 {
  font-size: 36px;
}

.h2SecondLine {
  font-size: 60px;
}

h3 {
  font-size: 24px;
  font-weight: normal;
  text-transform: uppercase;
  margin-top: 45px;
}

h4 {
  font-size: 24px;
  font-weight: normal;
  text-transform: uppercase;
}

hr {
  height: 1px;
  border: 0;
  border-top: 1px solid #c4c4c4;
  margin-top: 34px;
  margin-bottom: 34px;
}

.buttonLinkP {
  font-size: 14px;
  padding: 15px;
  text-align: center;
  border: 1px solid #212121;
  width: 100%;
  background-color: white;
}

.buttonLinkP a {
  text-transform: uppercase;
  text-decoration: none;
}

.alternateLinkP {
  font-size: 12px;
  text-align: center;
}

.alternateLinkP a {
  font-weight: normal;
}

.blackHighlight {
  background-color: #212121;
  color: white;
  padding: 20px;
  margin-top: 30px;
  margin-bottom: 30px;
}

.nobreak {
  white-space: nowrap;
}
