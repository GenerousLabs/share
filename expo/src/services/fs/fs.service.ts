import Bluebird from "bluebird";
import * as FileSystem from "expo-file-system";
import { groupBy } from "remeda";
import { superpathjoin } from "superpathjoin";
import { ENOENT, gitFsHttp, ALL_PATHS } from "../../shared.constants";

export const getDirectories = async () => {};

export const join = superpathjoin;

export const mkdirp = async ({ path }: { path: string }) => {
  // We can use FileSystem directly, which has its own `mkdir -p` equivalent
  await FileSystem.makeDirectoryAsync(
    join(FileSystem.documentDirectory, path),
    { intermediates: true }
  );
};

export const getDirectoryContents = async ({ path }: { path: string }) => {
  const { fs } = gitFsHttp;

  const names = await fs.promises.readdir(path);

  const entries = names.filter((name) => {
    // Ignore all filenames beginning with `.`
    if (name[0] === ".") {
      return false;
    }
    return true;
  });

  const stats = await Bluebird.map(entries, async (entry) => {
    const entryPath = join(path, entry);
    const stat = await fs.promises.stat(entryPath);
    return { name: entry, path: entryPath, stat };
  });

  const { directories = [], files = [] } = groupBy(stats, (entry) => {
    return entry.stat.isDirectory() ? "directories" : "files";
  });

  return { directories, files };
};

export const doesDirectoryExist = async ({ path }: { path: string }) => {
  const { fs } = gitFsHttp;

  return fs.promises.stat(path).then(
    (stat) => {
      return stat.isDirectory();
    },
    (error) => {
      if (error.code === ENOENT) {
        return false;
      }
      throw error;
    }
  );
};

export const doesFileExist = async ({ path }: { path: string }) => {
  const { fs } = gitFsHttp;

  return await fs.promises.stat(path).then(
    (stat) => {
      return stat.isFile();
    },
    (error) => {
      if (error.code === ENOENT) {
        return false;
      }
      throw error;
    }
  );
};

export const DANGEROUS_deleteEverything = async () => {
  await Bluebird.each(ALL_PATHS, async (path) => {
    await FileSystem.deleteAsync(join(FileSystem.documentDirectory, path), {
      idempotent: true,
    });
  });
};
