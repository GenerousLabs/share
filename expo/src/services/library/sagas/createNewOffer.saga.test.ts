import { describe, expect, it, jest } from "@jest/globals";
import fs from "expo-fs";
import { testSaga } from "redux-saga-test-plan";
import { RepoType } from "../../../shared.constants";
import { RepoInRedux } from "../../../shared.types";
import { createNewOfferSagaAction, effect } from "./createNewOffer.saga";

let mockStart = 1607000000000;
jest.mock("../../../utils/time.utils", () => {
  return {
    getTimestampSeconds: () => {
      const d = new Date(mockStart);
      // Add 10s to each call
      mockStart = mockStart + 10e3;
      return Math.round(d.getTime() / 1e3);
    },
  };
});

describe("createNewOfferEffect()", () => {
  const exampleRepo = {
    id: "repo1",
    name: "repo1",
    title: "repo1",
    bodyMarkdown: "",
    isReadOnly: false,
    remoteUrl: "",
    type: RepoType.library,
    headCommitObjectId: "",
    lastFetchTimestamp: 0,
  } as RepoInRedux;

  it("Creates a new offer #nsn5S8", () => {
    testSaga(
      effect,
      createNewOfferSagaAction({
        offer: {
          uuid: "offer1-uuid",
          bodyMarkdown: "An offer example",
          proximity: 0,
          shareToProximity: 1,
          title: "An example offer",
          tags: [],
        },
        repoId: "repo1",
      })
    )
      .next()
      .inspect((effect) => expect(effect).toMatchSnapshot())
      .next(exampleRepo)
      .call(fs.promises.mkdir, "/repos/repo1/an-example-offer")
      .next()
      .inspect((effect) => expect(effect).toMatchSnapshot())
      //       .call(
      //         fs.promises.writeFile,
      //         "/repos/repo1/an-example-offer/index.md",
      //         `---
      // uuid: offer1-uuid
      // proximity: 0
      // shareToProximity: 1
      // title: An example offer
      // tags: []
      // createdAt: 1607000000
      // updatedAt: 1607000000
      // ---
      // An offer example
      // `,
      //         { encoding: "utf8" }
      //       )
      .next()
      .inspect((effect) => expect(effect).toMatchSnapshot())
      .next()
      .inspect((effect) => expect(effect).toMatchSnapshot())
      .next()
      .isDone();
  });
});
