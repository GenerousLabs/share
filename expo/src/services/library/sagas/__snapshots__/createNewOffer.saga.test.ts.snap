// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createNewOfferEffect() Creates a new offer #nsn5S8 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "/repos/repo1/an-example-offer/index.md",
      "---
uuid: offer1-uuid
proximity: 0
shareToProximity: 1
title: An example offer
tags: []
createdAt: 1607000000
updatedAt: 1607000000
---
An offer example
",
      Object {
        "encoding": "utf8",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`createNewOfferEffect() Creates a new offer #nsn5S8 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      Object {
        "payload": Object {
          "message": "Creating a new offer",
          "repoId": "repo1",
        },
        "type": "SHARE/repo/commitAll",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`createNewOfferEffect() Terminates with an error for a non existent repoId #dussrU 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "meta": Object {
        "isFatal": false,
      },
      "payload": Object {
        "error": undefined,
        "message": "Repo does not exist #xJeqQd",
        "meta": Object {
          "repoId": "repo1",
        },
      },
      "type": "SHARE/library/createNewOffer/ERROR",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;
